include/deal.II/differentiation/ad/ad_number_traits.h

```diff
@@ -1060,58 +1082,365 @@ namespace Differentiation
-        return 0.0;
+            "Floating point / arithmetic numbers have no directional derivatives."));
+        return 0;
       }
+
+
+      static_assert((is_real_valued == true ?
+                       std::is_same<ad_type, real_type>::value :
+                       std::is_same<ad_type, complex_type>::value),
+                    "Incorrect template type selected for ad_type");
+
+      static_assert((is_complex_valued == true ?
+                       boost::is_complex<scalar_type>::value :
+                       true),
+                    "Expected a complex float_type");
+
+      static_assert((is_complex_valued == true ?
+                       boost::is_complex<ad_type>::value :
+                       true),
+                    "Expected a complex ad_type");
     };
 
+#  ifdef __clang__
+
+    template <typename ScalarType>
+    const bool NumberTraits<
+      ScalarType,
+      NumberTypes::none,
+      typename std::enable_if<
+        std::is_arithmetic<ScalarType>::value ||
+        (boost::is_complex<ScalarType>::value &&
+         std::is_arithmetic<typename internal::RemoveComplexWrapper<
+           ScalarType>::type>::value)>::type>::is_taped = false;
+
+
+    template <typename ScalarType>
+    const bool NumberTraits<
+      ScalarType,
+      NumberTypes::none,
+      typename std::enable_if<
+        std::is_arithmetic<ScalarType>::value ||
+        (boost::is_complex<ScalarType>::value &&
+         std::is_arithmetic<typename internal::RemoveComplexWrapper<
+           ScalarType>::type>::value)>::type>::is_tapeless = false;
+
+
+    template <typename ScalarType>
+    const bool NumberTraits<
+      ScalarType,
+      NumberTypes::none,
+      typename std::enable_if<
+        std::is_arithmetic<ScalarType>::value ||
+        (boost::is_complex<ScalarType>::value &&
+         std::is_arithmetic<typename internal::RemoveComplexWrapper<
+           ScalarType>::type>::value)>::type>::is_real_valued =
+      (!boost::is_complex<ScalarType>::value);
+
+
+    template <typename ScalarType>
+    const bool NumberTraits<
+      ScalarType,
+      NumberTypes::none,
+      typename std::enable_if<
+        std::is_arithmetic<ScalarType>::value ||
+        (boost::is_complex<ScalarType>::value &&
+         std::is_arithmetic<typename internal::RemoveComplexWrapper<
+           ScalarType>::type>::value)>::type>::is_complex_valued =
+      !(NumberTraits<ScalarType, NumberTypes::none>::is_real_valued);
+
+
+    template <typename ScalarType>
+    const unsigned int NumberTraits<
+      ScalarType,
+      NumberTypes::none,
+      typename std::enable_if<
+        std::is_arithmetic<ScalarType>::value ||
+        (boost::is_complex<ScalarType>::value &&
+         std::is_arithmetic<typename internal::RemoveComplexWrapper<
+           ScalarType>::type>::value)>::type>::n_supported_derivative_levels =
+      0;
+
+#  else
+
+    template <typename ScalarType>
+    constexpr bool NumberTraits<
```

Comment:
what are the rules for defining `constexpr` variables in header files? I forget...